test/AutoDiff/Sema/derivative_attr_type_checking.swift

```diff
@@ -802,3 +803,122 @@ extension HasADefaultDerivative {
+}
+
+func internal_original_public_derivative(_ x: Float) -> Float { x }
+// expected-error @+1 {{derivative function must have same access level as original function; derivative function '_internal_original_public_derivative' is public, but original function 'internal_original_public_derivative' is internal}}
+@derivative(of: internal_original_public_derivative)
+// expected-note @+1 {{mark the derivative function as 'internal' to match the original function}}
+public func _internal_original_public_derivative(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
+  fatalError()
+}
+
+private func private_original_usablefrominline_derivative(_ x: Float) -> Float { x }
+// expected-error @+1 {{derivative function must have same access level as original function; derivative function '_private_original_usablefrominline_derivative' is internal, but original function 'private_original_usablefrominline_derivative' is private}}
+@derivative(of: private_original_usablefrominline_derivative)
+@usableFromInline
+// expected-note @+1 {{mark the derivative function as 'private' to match the original function}}
+func _private_original_usablefrominline_derivative(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
+  fatalError()
+}
+
+private func private_original_public_derivative(_ x: Float) -> Float { x }
+// expected-error @+1 {{derivative function must have same access level as original function; derivative function '_private_original_public_derivative' is public, but original function 'private_original_public_derivative' is private}}
+@derivative(of: private_original_public_derivative)
+// expected-note @+1 {{mark the derivative function as 'private' to match the original function}}
+public func _private_original_public_derivative(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
+  fatalError()
+}
+
+private func private_original_internal_derivative(_ x: Float) -> Float { x }
+// expected-error @+1 {{derivative function must have same access level as original function; derivative function '_private_original_internal_derivative' is internal, but original function 'private_original_internal_derivative' is private}}
+@derivative(of: private_original_internal_derivative)
+// expected-note @+1 {{mark the derivative function as 'private' to match the original function}}
+func _private_original_internal_derivative(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
+  fatalError()
+}
+
+// MARK: - Original function visibility > derivative function visibility
+
+public func public_original_private_derivative(_ x: Float) -> Float { x }
+// expected-error @+1 {{derivative function must have same access level as original function; derivative function '_public_original_private_derivative' is fileprivate, but original function 'public_original_private_derivative' is public}}
+@derivative(of: public_original_private_derivative)
+// expected-note @+1 {{mark the derivative function as 'public' to match the original function}}
```

Comment:
Wait this is not a good idea. We should always suggest marking this function as `@usableFromInline internal`.

Reply:
Yes, thanks! Done.

The fix-it for "public original" simply suggests adding `@usableFromInline`. This fix-it works in at most two steps, even if the derivative access level is not internal.

Example:

```swift
import _Differentiation

public func publicOriginal(_ x: Float) -> Float { x }

@derivative(of: publicOriginal)
private func privativeDerivative(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
  fatalError()
}
```
```console
$ swift err.swift
err.swift:5:17: error: derivative function must have same access level as original function; derivative function 'privativeDerivative' is private, but original function 'publicOriginal' is public
@derivative(of: publicOriginal)
                ^
err.swift:6:14: note: mark the derivative function as '@usableFromInline' to match the original function
private func privativeDerivative(_ x: Float) -> (value: Float, pullback: (Float) -> Float) {
             ^
@usableFromInline
```

(The `^` position above seems weird. The source location is just `ValueDecl::getAttributeInsertionLoc(/*forModifier*/ false)`, I'm not sure why it's wrong and would prefer not to look further for now.)

After adding `@usableFromInline`:
```console 
$ swift err.swift
err.swift:5:1: error: '@usableFromInline' attribute can only be applied to internal declarations, but 'privativeDerivative' is private
@usableFromInline
^~~~~~~~~~~~~~~~~
```

After changing derivative access level to internal: `swift err.swift` succeeds.

Reply:
Looks good.